R/utils.R

```diff
@@ -167,11 +168,232 @@ match_filenames <- function(file_list, regex_list) {
+  required_vars <- sapply(schema$fields, function(x) {
+    if (!is.null(x$constraints) && !is.null(x$constraints$required)) {
+      return(x$constraints$required)
+    } else {
+      return(FALSE)
+    }
+  })
+  required_fields <- sapply(schema$fields[required_vars], function(x) x$name)
+  return(required_fields)
+}
+
+# Recursive function to extract variable names from frictionless schema
+
+#' @param lst the "fields" list of a frictionless schema
+#' @returns a character vector of variable names in the order of a VPTS CSV schema
+#' @keywords internal
+#' @noRd
+extract_names <- function(lst) {
+  if (is.list(lst)) {
+    names <- lapply(lst, function(x) extract_names(x$name))
+    unlist(names)
+  } else {
+    lst
+  }
+}
+
+#' Convert a tibble into a matrix
+#'
+#' Reshapes a tibble as an m✕n matrix of m distinct radar heights and
+#' n observations (sweeps) ordered by time. Each tibble contains data of a single vpts attribute.
+#' @param tibble A tibble in the format: datetime, height, variable, value.
+#' @return A list with two elements: the 'variable' of interest and the reshaped matrix
+#' @keywords internal
+#' @noRd
+tibble_to_mat <- function(tibble) {
+  unique_heights <- unique(tibble[["height"]])
+  matrix_list <- lapply(unique_heights, function(height) {
+    height_subset <- tibble[tibble[["height"]] == height, ]
+    matrix(height_subset$value, nrow = 1)
+  })
+  matrix <- do.call(rbind, matrix_list)
+  return(list(variable = tibble$variable[1], matrix = matrix))
+}
+
+#' Convert a vpts dataframe into an ordered list of matrices
+#'
+#' @param data A dataframe created from a VPTS CSV file
+#' @param maskvars a character vector of radar variables to be masked from the input , e.g., c("radar_latitude", "radar_longitude", ...)
+#' @param schema a frictionless schema
+#' @returns A named list of matrices ordered according to radvars
+#' @keywords internal
+#' @noRd
+df_to_mat_list <- function(data, maskvars, schema) {
+  datetime <- height <- variable <- fields <- NULL
+  radvars <- extract_names(schema$fields)
+  radvars <- radvars[!radvars %in% maskvars]
+
+  tbls_lst <- data %>%
+    dplyr::select(c(setdiff(colnames(data), maskvars), "datetime", "height")) %>%
+    tidyr::pivot_longer(-c(datetime, height), names_to = "variable", values_to = "value") %>%
+    dplyr::group_by(variable) %>%
+    dplyr::group_split()
+
+  unnamed_mat_list <- lapply(tbls_lst, tibble_to_mat)
+  var_names <- sapply(unnamed_mat_list, function(x) x$variable)
+  named_mat_list <- lapply(unnamed_mat_list, `[[`, "matrix")
+  names(named_mat_list) <- var_names
+  ordered_mat_list <- named_mat_list[radvars]
+
+  return(ordered_mat_list)
+}
+
+#' Convert a dataframe into a vpts object
+#'
+#' @param data a dataframe created from a VPTS CSV file
+#' @returns a bioRad vpts object
+#' @examples
+#' df = read.csv(system.file("extdata", "example_vpts.csv", package = "bioRad"))
+#' as.vpts(df)
+#' @export
+as.vpts <- function(data) {
```

Comment:
Each exported function should have its own file, so this should go into as.vpts.R